Add visual regression tests with playwright

martinRenou opened the pull request:
Fix https://github.com/spyder-ide/docrepr/issues/39

martinRenou committed:
Visual regression tests with playwright

martinRenou commented on .github/workflows/test.yaml:
```diff
@@ -21,7 +21,7 @@ jobs:
       matrix:
         os: [ubuntu-latest, windows-latest, macos-latest]
-        python-version: ['3.6', '3.10']
+        python-version: ['3.7', '3.10']
```
Playwright needs Python>=3.7

CAM-Gerlach replied:
While Python 3.6 just hit EoL, since we've preserved support up to the docrepr 0.2.0 release, I'd think it makes sense to release at least one version with it. Given it would be helpful to have the status of the playwright checks reported separately from the rest, and inter-Python-version differences that cause meaningful deltas in the output but don't fail any previous checks or raise an error (unsupported syntax, etc) seem very unlikely, I'd suggest just including a separate set of jobs in the matrix, one on each platform using e.g. Py3.8 that passes the custom pytest flag to enable the screenshot checks.

martinRenou commented:
@CAM-Gerlach this is a first draft of the visual regression testing. Please tell me what you think so far.

I am currently checking the diff between the actual screenshot and the expected result. We should probably check that diff with a possible percentage of error (there seems to be a slight difference in rendering fonts between Windows/Mac/Ubuntu).
Either we allow a percentage of error, or we store reference screenshots on an OS-basis: one reference directory per platform.

I am not sure which is the best, or if there is another more sensible approach?

CAM-Gerlach commented:
Thanks, looking great so far! Some general comments:

* [x] Since running the tests with playwright is somewhat non-trivial, could this be put behind a custom Pytest flag, e.g. `--compare-screenshots`, like `--open-browser` currently is (which should also be retained for manual testing and experimentation)? This also is needed to only run on a specific set of CI jobs. You could also enable the warning filters inside there.
* [x] To avoid encouraging a browser monoculture, could we use FF instead of chromium?
* [x] If we do deliberately change something with the UI, or add a new test, we'll need to generate new screenshots, which should hopefully be fairly low friction to avoid excessive maintenance overhead. As such, could you add a Python script, GH Actions job or some other automated, cross-platform mechanism to do so (or at least document step by step what needs to be done) .
* [x] ~Right now, it looks like the artifacts from different Python versions on the same platform are overwriting each other—you could include the Python version in the filename to fix that, but the above approach (just running it on one Python version) is probably better~
* [x] I suggest running optipng on the PNG output (CI-side, if its available); only takes a few seconds to run on all the images but saves 25-33% space for free

> I am currently checking the diff between the actual screenshot and the expected result. We should probably check that diff with a possible percentage of error (there seems to be a slight difference in rendering fonts between Windows/Mac/Ubuntu).
Either we allow a percentage of error, or we store reference screenshots on an OS-basis: one reference directory per platform.

I'd think per-OS screenshots, if you can manage that, since we'd need to pick and test a threshold very carefully to avoid false negatives, and there's the risk that it could miss significant changes.

martinRenou commented:
Thanks for the review, those are all very good points I'll finish this on Monday 😊

martinRenou commented:
> I suggest running optipng on the PNG output (CI-side, if its availible); only takes a few seconds to run on all the images but saves 25-33% space for free

Are you suggesting this for the build artifacts? Or would you like this to be done for references that are stored in the repo as well?

martinRenou commented:
~Trying the action (this might not work as it's not part of master yet):~

~Please generate reference screenshots~

That does not work

CAM-Gerlach commented:
FYI, I mean to suggest this when I replied before, but you might want to remove ` --maxfail 5` in `addopts` in `pytest.ini`—then you'll get the full output for all tests when they fail, not just the first five. I shouldn't have added it in the first place, that line was a copy-paste from another repo I set up with it that needed it for some reason.

> Are you suggesting this for the build artifacts? Or would you like this to be done for references that are stored in the repo as well?

Ideally for both, but most importantly for the reference screenshots, since they will be stored in the repo and each version will take up space for all time, so the smaller we can get them, the better. Note that this should be done as part of the GitHub workflow that generates them, rather than inside the test code itself.

CAM-Gerlach commented:
> Trying the action (this might not work as it's not part of master yet):

Yeah, I'm sure that's what happened. We can do a best effort on this PR, merge it and then test it and follow up if we need to fix anything with that part. Or, you can merge it on your fork and test it there,, since that _should_ work.

BTW, you can get rid of all build and install steps and replace all that with just `pip install --upgrade .`; also, it only needs to run on Py3.8 that the visual regression tests themselves will run on.

Let me know if you need anything!

martinRenou commented:
> FYI, I mean to suggest this when I replied before, but you might want to remove --maxfail 5 in addopts in pytest.ini—then you'll get the full output for all tests when they fail, not just the first five. I shouldn't have added it in the first place, that line was a copy-paste from another repo I set up with it that needed it for some reason.

Right, thanks! I was a bit confused why it was behaving this way but didn't look into it.

> Note that this should be done as part of the GitHub workflow that generates them, rather than inside the test code itself.

👍🏽 

> Or, you can merge it on your fork and test it there

Yes, I was doing this yesterday and it does not work properly. I'm looking into it.

martinRenou commented:
I am getting there: https://github.com/martinRenou/docrepr/pull/8

Commenting "Please generate reference screenshots" in the PR properly triggers a commit of the references :) I still need to polish a bit and fix some remaining issues, but this is taking shape

martinRenou commented on .github/workflows/test.yaml:
```diff
@@ -52,3 +52,41 @@ jobs:
+      run: python -m playwright install firefox
+    - name: Run tests
+      shell: bash
+      run: python -I -bb -X dev -W error -m pytest --compare-screenshots --asyncio-mode=auto
```
We can move the `asyncio-mode` option to the `pytest.ini` file once we remove Python 3.6 tests. This option makes Python 3.6 tests fail because `pytest-asyncio<=0.16` (latest version supporting Python 3.6) does not implement this option.

martinRenou commented:
I resolved all of your comments :)

As you can see from the references, we can reproduce this issue https://github.com/spyder-ide/docrepr/issues/38. I'll try to fix it in a separate PR.

CAM-Gerlach commented on .github/workflows/test.yaml:
```diff
@@ -46,9 +46,52 @@ jobs:
+      if: ${{ matrix.python-version == '3.6' }}
       shell: bash
       run: python -I -bb -X dev -W error -m pytest
+    - name: Run tests
```
To be more DRY instead of duplicating these two steps with one additional argument, I suggest just adding a `pytest-args` variable to the matrix for `python-version: '3.6'` via `include`, and injecting that at the end of the `run` command. See what I did with `special-invocation` in the [QtPy CI workflow](https://github.com/spyder-ide/qtpy/blob/master/.github/workflows/ci.yml) for an example (except in that case, I injected it in the default `shell` instead of the step-specific `run`).

CAM-Gerlach commented on .github/workflows/test.yaml:
```diff
@@ -46,9 +46,52 @@ jobs:
     - name: Pip check
       run: pip check -v
+
+  visual-regression-tests:
```
```suggestion
  visual-regression-test:
```

Since aside from trigger conditions this doesn't share anything with the other workflow, I suggest just sticking this in a separate workflow file so it can be rerun separately from the regular tests as needed (which is a significant pain point with GitHub actions). Also, we should be consistent about naming; either call this `visual-regression-test`, or rename the other job `test` (and same with the extras names).

CAM-Gerlach commented on .github/workflows/update_references.yaml:
```diff
@@ -0,0 +1,56 @@
+
+jobs:
+  update_references:
+
```
```suggestion
```

Stray line break

CAM-Gerlach commented on .github/workflows/update_references.yaml:
```diff
@@ -0,0 +1,56 @@
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest]
+    steps:
+
```
```suggestion
```

Another stray line break

CAM-Gerlach commented on .github/workflows/update_references.yaml:
```diff
@@ -0,0 +1,56 @@
+    types: [created, edited]
+
+
+jobs:
```
```suggestion
jobs:
  strategy:
    max_parallel: 1
```

To avoid race conditions with pulling and committing the screenshots causing the job to fail

CAM-Gerlach commented on .github/workflows/update_references.yaml:
```diff
@@ -0,0 +1,56 @@
+          git config user.name 'github-actions[bot]'
+          git config user.email 'github-actions[bot]@users.noreply.github.com'
+
+          git pull
+
+          git add docrepr/tests/references
+
+          git commit -m "Update references for ${{ matrix.os }}"
+          git push
```
There's a fair chance you could randomly hit a race condition here, based on the timing of the jobs, if the jobs execute close enough together that one job pulls, another job pulls and the first job pushes. To avoid this, I set `max_parallel: 1` above.

martinRenou replied:
Thanks!

CAM-Gerlach commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,8 @@
+## Visual changes
+
+If this PR has an impact on the outputs, you can ask the bot to update
+the references for the visual regression tests commenting the following:
```
```suggestion
If this PR has an impact on the rendered output, you can ask the bot to
update the reference screenshots for the visual regression tests by
commenting the following:
```

Fix a few minor grammar/phrasing issues

CAM-Gerlach commented on .gitignore:
```diff
@@ -72,6 +72,8 @@ pip-delete-this-directory.txt
 htmlcov/
 nosetests.xml
+docrepr/tests/screenshots/**
+docrepr/tests/diffs/**
```
```suggestion
docrepr/tests/screenshots/
docrepr/tests/diffs/
```

Shouldn't be needed, as ending with a `/` should ignore the whole dir, like the others listed

CAM-Gerlach commented on setup.cfg:
```diff
@@ -62,3 +62,6 @@ test =
     numpy
     pytest>=6.0.0
+    pytest-asyncio
+visual_tests =
```
```suggestion
visual_test =
```

To be consistent with the tests `test` extra

CAM-Gerlach commented on .github/workflows/test.yaml:
```diff
@@ -46,9 +46,52 @@ jobs:
+        import sys
+        print(sys.executable, sys.version, sep='\n')
+    - name: Install docrepr
+      run: python -m pip install --upgrade .[test,visual_tests]
```
```suggestion
      run: python -m pip install --upgrade .[test,visual_test]
```

CAM-Gerlach commented on .github/workflows/update_references.yaml:
```diff
@@ -0,0 +1,56 @@
+          import sys
+          print(sys.executable, sys.version, sep='\n')
+      - name: Install docrepr
+        run: python -m pip install --upgrade .[test,visual_tests]
```
```suggestion
        run: python -m pip install --upgrade .[test,visual_test]
```

CAM-Gerlach commented on .github/workflows/update_references.yaml:
```diff
@@ -0,0 +1,56 @@
+        run: pip list
+      - name: Install Firefox
+        run: python -m playwright install firefox
+      - name: Run tests
```
```suggestion
      - name: Generate updated reference screenshots
```

CAM-Gerlach commented on .github/workflows/update_references.yaml:
```diff
@@ -0,0 +1,56 @@
+jobs:
+  update_references:
+
+    name: PR comment
```
```suggestion
    name: Update reference screenshots on ${{ matrix.os }}
```

Makes a little more sense here, no?

CAM-Gerlach commented on .github/workflows/test.yaml:
```diff
@@ -46,9 +46,52 @@ jobs:
+    - name: Run tests
+      shell: bash
+      run: python -I -bb -X dev -W error -m pytest --compare-screenshots --asyncio-mode=auto
+    - name: Upload UI Test artifacts
```
```suggestion
    - name: Upload UI test artifacts
```

Consistent sentence case

CAM-Gerlach commented on conftest.py:
```diff
@@ -1,28 +1,45 @@
+import shutil
+from sys import platform
 import warnings
 import webbrowser
```
```suggestion
import shutil
import sys
import warnings
import webbrowser
from pathlib import Path
```

`from` below `import` is the most standard style AFAIK and the `isort` default, and what we use in the other modules.

CAM-Gerlach commented on conftest.py:
```diff
@@ -1,28 +1,45 @@
+        COMPARE_SCREENSHOTS_OPTION,
+        action='store_true',
+        default=False,
+        help='For tests that generate HTML output, run visual regression tests on them',
```
```suggestion
        help=('For tests that generate HTML output, '
              'run visual regression tests on it'),
```

Plural agreement and 79 chars

CAM-Gerlach commented on conftest.py:
```diff
@@ -1,28 +1,45 @@
+        UPDATE_REFERENCE_SCREENSHOTS_OPTION,
+        action='store_true',
+        default=False,
+        help='For tests that generate HTML output, update reference screenshots for the visual regression tests',
```
```suggestion
        help=('For tests that generate HTML output, '
              'update reference screenshots for the visual regression tests'),
```

CAM-Gerlach commented on conftest.py:
```diff
@@ -37,3 +54,59 @@ def _open_browser(url):
+            request.config.getoption(UPDATE_REFERENCE_SCREENSHOTS_OPTION)):
+            from playwright.async_api import async_playwright
+
+            # Filtering warnings generated by playwright
```
```suggestion
            # Filter warnings generated by playwright
```

Imperative tense

CAM-Gerlach commented on conftest.py:
```diff
@@ -37,3 +54,59 @@ def _open_browser(url):
+            warnings.filterwarnings(
+                'ignore', category=DeprecationWarning, module='pyee.*')
+
+            test_dir = Path(__file__).parent / 'docrepr' / 'tests'
```
Not great that we have to hardcode this path here, but you put the fixture in a `conftest` in the test directory or directly in `test_output`, you'd have to hardcod the option names instead, so I guess this is the least bad approach. Maybe at least factor this into a top-level constant, `TEST_DIR`, so if it does change, we can change it one obvious place?

CAM-Gerlach commented on conftest.py:
```diff
@@ -37,3 +54,59 @@ def _open_browser(url):
+                'ignore', category=DeprecationWarning, module='pyee.*')
+
+            test_dir = Path(__file__).parent / 'docrepr' / 'tests'
+            image = f'test-{test_id}-{platform}.png'
```
```suggestion
            image = f'test-{test_id}-{sys.platform}.png'
```

To match above change

CAM-Gerlach commented on conftest.py:
```diff
@@ -1,28 +1,45 @@
 # ---- Constants
 
 OPEN_BROWSER_OPTION = '--open-browser'
+COMPARE_SCREENSHOTS_OPTION = "--compare-screenshots"
+UPDATE_REFERENCE_SCREENSHOTS_OPTION = "--update-reference-screenshots"
```
```suggestion
COMPARE_SCREENSHOTS_OPTION = '--compare-screenshots'
UPDATE_REFERENCE_SCREENSHOTS_OPTION = '--update-reference-screenshots'
```

Consistent string quotes (I normalized the handful that were inconsistent in #40 )

CAM-Gerlach commented on conftest.py:
```diff
@@ -37,3 +54,59 @@ def _open_browser(url):
+            screenshot = (test_dir / 'screenshots' / image).resolve()
+            diff = (test_dir / 'diffs' / image).resolve()
+
+            # Create diff directory
+            (test_dir / 'diffs').mkdir(parents=True, exist_ok=True)
```
Should we also create the screenshots directory here, just for consistency? (I'm assuming it gets created implicitly by the `page.screenshot()` call...)

CAM-Gerlach commented on conftest.py:
```diff
@@ -37,3 +54,59 @@ def _open_browser(url):
+            async with async_playwright() as p:
+                browser = await p.firefox.launch()
+                page = await browser.new_page()
+                await page.goto(f'file://{url}', wait_until="networkidle")
```
```suggestion
                await page.goto(f'file://{url}', wait_until='networkidle')
```

Consistancy

CAM-Gerlach commented on docrepr/tests/test_output.py:
```diff
@@ -4,6 +4,7 @@
 import copy
 import subprocess
 import sys
+from time import sleep
```
```suggestion
```

I think this is a leftover?

CAM-Gerlach commented on docrepr/tests/test_output.py:
```diff
@@ -166,7 +167,10 @@ def eat_one(self):
 def _test_cases_to_params(test_cases):
-    return [tuple(test_case.values()) for test_case in test_cases.values()]
+    return [
+        [test_id, *test_case.values()]
```
```suggestion
        (test_id, *test_case.values())
```

Is there a reason this was changed from a tuple (per general semantics and Pytest convention) to a list?

CAM-Gerlach reviewed (changes requested):
Thanks so much for all your hard work on this, @martinRenou ! This looks really great. I just have some relatively minor comments, and then this should be good to go.

Top level comments:

* [x] I suggest running `optipng` on the reference PNGs output by the CI (as well as those committed here); since they will be committed irrevocably to the repo once merged, and will inexorably increase its size, the smaller we can make them, the better.
* [x] To be more clear, consistent and descriptive about what we're referring to, follow idiomatic English convention and avoid confusion with more typical meanings of "references" in English ("bibliography", people that will vouch for your character, etc), could you rename all instances of `references` to `reference_screenshots` or similar (as appropriate)?

To note, I'm not super comfortable with asyncio beyond the very basics (since we don't really use it in my main lines of work) and this doesn't seem to be too performance-critical given the complexity, but I suppose this is acceptable since it is only in the tests and mostly isolated to the one fixture...so long as you're willing to help fix any async-related issues that may come up.

Thanks again!

martinRenou commented:
> I suggest running optipng on the reference PNGs output by the CI (as well as those committed here); since they will be committed irrevocably to the repo once merged, and will inexorably increase its size, the smaller we can make them, the better.

I've implemented the compression with Pillow instead: https://github.com/spyder-ide/docrepr/pull/41/files#diff-a31c7ed5d35f5ed8233994868c54d625b18e6bacb6794344c4531e62bd9dde59R97-R99 this logic also works for the reference screenshots (it has been used for the references on this PR).

I am not sure optipng can be easily installed on all platforms? Maybe Pillow's compression should be enough? I can decrease the quality even more if you want.

> To note, I'm not super comfortable with asyncio beyond the very basics (since we don't really use it in my main lines of work) and this doesn't seem to be too performance-critical given the complexity, but I suppose this is acceptable since it is only in the tests and mostly isolated to the one fixture...so long as you're willing to help fix any async-related issues that may come up.

At first I implemented it without asyncio, as playwright provides a sync API: https://github.com/microsoft/playwright-python#example

The issue was that the sync API was not cleaning up resources and shutting down the browser properly (it was generating many warnings that needed to be filtered). We can roll-back to the sync API but we'll have to add filters. Also there was a warning that I was not able to filter for some reason, that's why I gave up and used the async API.

martinRenou commented:
Thanks for the review!

CAM-Gerlach commented:
Thanks for the quick and through response!

> I've implemented the compression with Pillow instead: https://github.com/spyder-ide/docrepr/pull/41/files#diff-a31c7ed5d35f5ed8233994868c54d625b18e6bacb6794344c4531e62bd9dde59R97-R99 this logic also works for the reference screenshots (it has been used for the references on this PR).

Yep, I'd noticed that—thanks. Unfortunately, Pillow's optimization still appears to be leaving a lot on the table, assuming its what you're using for the current reference screenshots—running `optipng` on them, even at the default (relatively low) optimization level, still results in between 10% a and 50% decrease in the size of each image, which reduces the total size of the screenshot directory from 1.5 MB to 1.1 MB, or between a 25% and 30% savings. That's not trivial considering the repo will grow in size by nearly that much every time we regenerate them, assuming they all change (consistent optimization might or might not reduce those deltas).

It actually shouldn't be that tough to install optipng on all GHA platforms—on Windows, its just `choco install optipng`, on Mac its `brew install optipng` and on Ubuntu, `apt install optipng`, then run it with `optipng docrepr/tests/reference_screenshots/*.png`. That's probably the simplest approach, but you could also upload the archives from each job, then have another job that runs after all three jobs have completed that merges, optimizes and commits them. This allows the three jobs to run in parallel, make a single commit at the end and not have to install optipng on each platform, but probably isn't worth the work unless you really want to.

> The issue was that the sync API was not cleaning up resources and shutting down the browser properly (it was generating many warnings that needed to be filtered). We can roll-back to the sync API but we'll have to add filters. Also there was a warning that I was not able to filter for some reason, that's why I gave up and used the async API.

Makes sense, thanks :+1: I saw you used the non-async API before, but I figured there was a reason you switched (and noticed most of the warnings were gone).

martinRenou commented:
Just added `optipng` to the `update-reference-screenshots` CI job :) thanks for your patience!

You can see that it runs properly here: 
https://github.com/martinRenou/docrepr/runs/4963763427?check_suite_focus=true

Which was triggered by https://github.com/martinRenou/docrepr/pull/24

CAM-Gerlach commented:
@martinRenou LGTM, thanks! Now that #40 is merged, the output should change for a number of tests; to avoid an extra copy of the images and them getting out of sync, do you mind rebasing this and using the CIs on your fork to generate the updated reference screenshots for each OS (make sure you use the CI-optimized ones, since different `optipng` builds can produce slightly different results, I've found)? Once that's done, I'll go ahead and merge this. Thanks!

martinRenou commented:
Rebased! :)

CAM-Gerlach reviewed (approved):
LGTM, thanks @martinRenou !
